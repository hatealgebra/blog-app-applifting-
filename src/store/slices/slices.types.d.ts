export enum ESortByOptions {
  ORIGINAL = 'original',
  BY_TITLE = 'title',
  BY_PEREX = 'perex',
  BY_AUTHOR = 'author',
  BY_NR_COMMENTS = 'comments',
}

export enum Hello {}
