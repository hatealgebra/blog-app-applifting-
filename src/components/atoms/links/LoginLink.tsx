import React from 'react';

import { HiOutlineArrowLeft } from '@react-icons/all-files/hi/HiOutlineArrowLeft';

import { StyledLoginLink } from './link.styled';
import { NavLinks } from '../../../utils/contants';

const LoginLink = () => {
  const { LOGIN } = NavLinks;
  return (
    <StyledLoginLink to={LOGIN} className="login-link">
      <span>Login</span>
      <HiOutlineArrowLeft style={{ transform: 'rotate(180deg)' }} />
    </StyledLoginLink>
  );
};

export default LoginLink;
