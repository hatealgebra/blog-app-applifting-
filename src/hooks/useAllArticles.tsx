import { useStaticQuery, graphql } from 'gatsby';
import { useEffect, useState } from 'react';

const useAllArticles = () => {
  const [articles, setArticles] = useState([]);

  const allArticles = useStaticQuery(graphql`
    query {
      allPosts {
        nodes {
          id
          articleId
          createdAt
          lastUpdatedAt
          imageId
          imageBase64
          title
          perex
          content
          comments
        }
      }
    }
  `);

  useEffect(() => {
    setArticles(allArticles.allPosts.nodes);
  }, []);

  return articles;
};

export default useAllArticles;
