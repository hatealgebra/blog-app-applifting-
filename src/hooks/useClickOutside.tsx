import React from 'react';

const useClickOutside = (
  ref: React.RefObject<Element>,
  setIsMenuOpen: React.Dispatch<React.SetStateAction<boolean>>
) => {
  React.useEffect(() => {
    const handleClickOutside = (e: MouseEvent) => {
      if (!ref.current || ref.current.contains(e.target)) {
        console.log(e.target);
      } else {
        setIsMenuOpen(false);
      }
    };

    document.addEventListener('mousedown', (e) => handleClickOutside(e));

    return () => {
      document.removeEventListener('mousedown', handleClickOutside);
    };
  }, [ref, setIsMenuOpen]);
};

export default useClickOutside;
