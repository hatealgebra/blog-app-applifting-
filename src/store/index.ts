import { configureStore, combineReducers } from '@reduxjs/toolkit';

import { persistStore, persistReducer } from 'redux-persist';
import storage from 'redux-persist/lib/storage';
import authReducer from './slices/auth.slices';
import articleFeedReducer from './slices/articleFeed.slice';
import adminReducer from './slices/admin.slices';

// REDUX-PERSIST

export const persistConfig = {
  key: 'root',
  storage,
};

export const persistedReducer = persistReducer(persistConfig, authReducer);

export const reducer = combineReducers({
  // FIXME: Just store the data where it is needed, use multiple stores
  articleFeed: articleFeedReducer,
  admin: adminReducer,
});

const store = configureStore({
  reducer: { reducer, persistedReducer },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: false,
    }),
  devTools: true,
});

export const persistor = persistStore(store);
export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export default store;
