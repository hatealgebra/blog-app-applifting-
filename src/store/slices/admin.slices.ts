/* eslint-disable no-param-reassign */
import { createSlice } from '@reduxjs/toolkit';

import { Components } from '../../customTypes/declarations';
import type { RootState } from '..';
import { getArticlesFeedThunk } from '../thunks/articles.thunk';
import { createArticleThunk, deleteArticleThunk } from '../thunks/admin.thunks';
import { ESortByOptions } from './slices.types.d';

const initialState = {
  status: 'idle',
  data: {},
  error: false,
} as {
  status: 'idle' | 'loading';
  error: boolean;
  data: {
    originalSort: Components['schemas']['ArticleList'];
    nowSort: Components['schemas']['ArticleList'];
    articleToEdit: Components['schemas']['ArticleDetail'];
  };
};

export const adminSlice = createSlice({
  name: 'admin',
  initialState,
  reducers: {
    sortMyArticles: (state, { payload }: { payload: ESortByOptions }) => {
      if (payload === ESortByOptions.ORIGINAL) {
        state.data.nowSort = state.data.originalSort;
      } else {
        state.data.nowSort.items = [...state.data.originalSort.items!].sort(
          (a, b) => {
            if (a[payload] > b[payload]) {
              return 1;
            }
            if (a[payload] < b[payload]) {
              return -1;
            }
            return 0;
          }
        );
      }
    },
    setArticleToEdit: (state, { payload }) => {
      console.log('payload', payload);
      state.data.articleToEdit = payload;
    },
  },
  extraReducers: (builder) => {
    builder.addCase(getArticlesFeedThunk.pending, (state) => {
      state.status = 'loading';
    });
    builder.addCase(getArticlesFeedThunk.fulfilled, (state, { payload }) => {
      state.status = 'idle';
      state.data.originalSort = payload;
      state.data.nowSort = payload;
    });
    builder.addCase(getArticlesFeedThunk.rejected, (state) => {
      state.error = true;
      state.status = 'idle';
      state.data = initialState.data;
    });

    builder.addCase(createArticleThunk.pending, (state) => {
      state.status = 'loading';
    });
    builder.addCase(createArticleThunk.fulfilled, (state, { payload }) => {
      state.data.originalSort?.items?.push(payload);
      state.status = 'idle';
    });
    builder.addCase(createArticleThunk.rejected, (state) => {
      state.status = 'idle';
    });

    builder.addCase(deleteArticleThunk.pending, (state) => {
      state.status = 'loading';
    });
    builder.addCase(deleteArticleThunk.fulfilled, (state, { payload }) => {
      state.data.originalSort.items = payload;
      state.status = 'idle';
    });
    builder.addCase(deleteArticleThunk.rejected, (state) => {
      state.status = 'idle';
    });
  },
});

// ACTIONS
export const { sortMyArticles, setArticleToEdit } = adminSlice.actions;
// SELECTORS
export const selectMyArticlesOriginalItems = (state: RootState) =>
  state.reducer?.admin.data.originalSort?.items;
export const selectMyArticlesItems = (state: RootState) =>
  state.reducer?.admin.data.nowSort?.items ?? [];
export const selectMyArticlesStatus = (state: RootState) =>
  state.reducer?.admin.status;
export const selectMyArticlesError = (state: RootState) =>
  state.reducer?.admin.error;
export const selectArticleToEdit = (state: RootState) =>
  state.reducer?.admin.data.articleToEdit;

export default adminSlice.reducer;
