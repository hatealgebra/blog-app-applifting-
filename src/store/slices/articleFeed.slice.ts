/* eslint-disable no-param-reassign */
import { createSlice } from '@reduxjs/toolkit';

import { Components } from '../../customTypes/declarations';
import type { RootState } from '..';
import { getArticlesFeedThunk } from '../thunks/articles.thunk';

const initialState = {
  status: 'idle',
  data: { items: undefined },
  error: false,
} as {
  status: 'idle' | 'loading';
  error: boolean;
  data: Components['schemas']['ArticleList'];
};

export const articleFeedSlice = createSlice({
  name: 'articleFeed',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(getArticlesFeedThunk.pending, (state) => {
      state.status = 'loading';
    });
    builder.addCase(getArticlesFeedThunk.fulfilled, (state, { payload }) => {
      state.status = 'idle';
      state.data = payload;
    });
    builder.addCase(getArticlesFeedThunk.rejected, (state) => {
      state.error = true;
      state.status = 'idle';
      state.data = initialState.data;
    });
  },
});

export const selectArticleFeedItems = (state: RootState) =>
  state.reducer.articleFeed.data.items;
export const selectArticleFeedStatus = (state: RootState) =>
  state.reducer.articleFeed.status;
export const selectArticleFeedError = (state: RootState) =>
  state.reducer.articleFeed.error;

export enum ArticleFeedActions {
  GET_ARTICLES = 'articleFeed/getArticlesFeedThunk',
}

export default articleFeedSlice.reducer;
