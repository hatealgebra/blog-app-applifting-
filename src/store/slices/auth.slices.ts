import { createSlice } from "@reduxjs/toolkit";
import { components } from "../../types/declarations";
import { RootState } from "..";

import {
  postLoginThunk,
  reAuthorizeThunk,
} from "../thunks/authentication.thunks";

export const authInitialState = {
  status: "idle",
  data: {},
  error: false,
} as {
  status: "idle" | "loading";
  data: {
    tenant?: components["schemas"]["Tenant"] | null;
    authorization?: components["schemas"]["AccessToken"];
    login?: {
      email: string;
      pwd: string;
    };
  };
  error: boolean;
};

export const authSlice = createSlice({
  name: "auth",
  initialState: authInitialState,
  reducers: {
    logout: (state) => {
      state.data = {};
    },
  },
  extraReducers: (builder) => {
    // login

    builder.addCase(postLoginThunk.pending, (state) => {
      state.status = "loading";
    });
    builder.addCase(postLoginThunk.fulfilled, (state, { payload }) => {
      state.status = "idle";
      state.data = payload;
    });
    builder.addCase(postLoginThunk.rejected, (state, { payload }) => {
      state.status = "idle";
      state.error = true;
    });
    // reauthorize
    builder.addCase(reAuthorizeThunk.pending, (state) => {
      state.status = "loading";
    });
    builder.addCase(reAuthorizeThunk.fulfilled, (state, { payload }) => {
      state.status = "idle";
      state.data.authorization = payload.data;
    });
    builder.addCase(reAuthorizeThunk.rejected, (state, { payload }) => {
      state.status = "idle";
      state.error = true;
    });
  },
});

export const { logout } = authSlice.actions;

export const logoutAction = { type: logout };

export const selectAuthName = (state: RootState) =>
  state.persistedReducer?.data.login?.email;
export const selectAuthTenant = (state: RootState) =>
  state.persistedReducer?.data.tenant;
export const selectAuthToken = (state: RootState) =>
  state.persistedReducer?.data.authorization?.access_token;
export const selectAuthStatus = (state: RootState) =>
  state.persistedReducer.status;
export const selectAuthError = (state: RootState) =>
  state.persistedReducer.error;
export const selectAuthLogged = (state: RootState) => {
  const authorization = state.persistedReducer?.data.authorization;
  const login = state.persistedReducer?.data.login;
  return !authorization || Object.keys(authorization!).length === 0 || !login
    ? false
    : true;
};

export default authSlice.reducer;
