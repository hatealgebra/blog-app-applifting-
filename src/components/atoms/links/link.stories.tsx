import React from "react";
import { useAppSelector } from "../../../store/hooks";
import { selectAuthToken } from "../../../store/slices/auth.slices";
import { clearDataAPI } from "../../../utils/generic.utils";
import Link from "./link.styled";
import LoginLink from "./LoginLink";

export const AllVariants = () => (
  <div style={{ display: "flex", flexDirection: "column", gap: "10px" }}>
    <Link to="/#">Classic link</Link>
    <Link variant="text" to="/#">
      Text link
    </Link>
    <LoginLink />
  </div>
);

export default {
  component: Link,
  subcomponent: { LoginLink },
  title: "Atoms/Link",
};
