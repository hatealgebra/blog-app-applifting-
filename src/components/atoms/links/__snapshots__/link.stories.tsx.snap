// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Atoms/Link AllVariants smoke-test 1`] = `
<div style="display: flex; flex-direction: column; gap: 10px;">
  <a class="sc-jrkPcm eEbEbs"
     href="/#"
  >
    Classic link
  </a>
  <a variant="text"
     class="sc-jrkPcm eEbEbs"
     href="/#"
  >
    Text link
  </a>
  <a class="sc-ipMvLY hcUcbw login-link"
     href="/login"
  >
    <span>
      Login
    </span>
    <svg stroke="currentColor"
         fill="none"
         stroke-width="0"
         viewbox="0 0 24 24"
         height="1em"
         width="1em"
         xmlns="http://www.w3.org/2000/svg"
         style="transform: rotate(180deg);"
    >
      <path stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            d="M10 19l-7-7m0 0l7-7m-7 7h18"
      >
      </path>
    </svg>
  </a>
</div>
`;
